src/examples/learn_quant/monotonicity.py

```diff
@@ -0,0 +1,254 @@
+    objects = names_vector.tolist()
+    properties = truth_df.columns
+    bools = switch_direction(truth_df, down)
+    return Context(objects, properties, bools)
+
+
+def get_sub_structures(concept_lattice: Context, name: list[str]) -> set[str]:
+    """Accepts a list of a singleton name to get a set of substructures of the indexed model, excluding the indexing model name.
+
+    Returns:
+        set(str): Tuple of substructures of the model indexed that exist in the universe.
+
+    """
+
+    return set(concept_lattice[name][0]) - set(name)
+
+
+def has_sub_structure(concept_lattice: Context, name: list[str]) -> bool:
+    """Identity function for a model having a substructure in the universe.
+
+    Returns:
+        bool: ``True`` if there is one or more substructures in the universe, otherwise ``False``.
+    """
+
+    return True if len(set(concept_lattice[name][0]) - set(name)) > 0 else False
+
+
+def get_sub_structure_in_meaning(
+    concept_lattice, name: list[str], meaning: Meaning
+) -> set[str]:
+    """Identity function for a model having a substructure in a defined Meaning.
+
+    Returns:
+        set(str): Tuple of substructures of the model indexed that exist in the selected Meaning.
+    """
+
+    names = set(referent.name for referent in meaning.referents)
+
+    return get_sub_structures(concept_lattice, name) & names
+
+
+def has_sub_structure_in_meaning(
+    concept_lattice, name: list[str], meaning: Meaning
+) -> bool:
+    """Identity function for a model having a substructure in a defined Meaning.
+
+    Returns:
+        bool: ``True`` if there is one or more substructures in the universe, otherwise ``False``.
+    """
+
+    return (
+        True
+        if len(get_sub_structure_in_meaning(concept_lattice, name, meaning)) > 0
+        else False
+    )
+
+
+def upward_monotonicity_entropy(all_models, quantifier):
+    """Measures degree of upward monotonicity of a quantifiers as
+    1 - H(Q | true_pred) / H(Q) where H is (conditional) entropy, and true_pred is the
+    variable over models saying whether there's a true _predecessor_ in the
+    subset order.
+
+    :param all_models: calculated matrix MxM, where a given row is an array of bools that correspond to whether a model at that index is a submodel
+    :param quantifier: list of truth values, same len as models
+    :returns scalar
+    """
+
+    submembership = all_models
+
+    if np.all(quantifier) or not np.any(quantifier):
+        return 1
+    p_q_true = sum(quantifier) / len(quantifier)
+    p_q_false = 1 - p_q_true
+    q_ent = -p_q_true * np.log2(p_q_true) - p_q_false * np.log2(p_q_false)
+
+    def get_preds(num_arr, num):
+        """Given an array of ints, and an int, get all predecessors of the
+        model corresponding to int.
+        Returns an array of same shape as num_arr, but with bools
+        """
+        return num_arr[num, :]
+
+    def has_true_pred(num_arr, y):
+        return np.any(y * num_arr)
```

Comment:
Is it true that these `get_preds` and `has_true_pred` are no longer needed, i.e. replaced by your lattice-based methods?

Reply:
Yeah, but I do wonder if we should not do the lattice methods because they might be slower than the logic you've written, I'm not yet sure

Reply:
The one big issue I foresee is that the methods I have here rely on very strong assumptions about the nature of the quantifiermodels that are no longer true (but were true in Carcassi et al 2021): we're assuming all models _are the same size_ and that there are only two sets ($A \cap B$ and $A \setminus B$).  This let us represent models by binary vectors of a fixed length.  Your "names" are similar, except now they're 5-ary vectors instead of binary vectors; I'd have to sit down and think whetehr the logic here can be used with those or if it needs to be modified (and, if so, whether that's simple to do or not)